Merge Cut by destination 2 to Main

Erabior opened the pull request.

Erabior committed:
uncoupling modes

Erabior committed:
test

Erabior committed:
merge main to Cut by Destination

Erabior committed:
fix merge errors

Erabior committed:
Bug fixes

Erabior committed:
Better destination selection working

Erabior committed:
bug fixes

Erabior committed:
revert

Erabior committed:
cleanup

Erabior committed:
Merge branch 'main' into Cut-by-destination

Erabior committed:
Merge branch 'Cut-by-destination' of https://github.com/PrinceOfPluto/WaypointQueue into Cut-by-destination

Erabior committed:
fixed issue where cut by destination was not able to find waybills in the consist

Erabior committed:
Merge branch 'main' into Cut-by-destination

Erabior committed:
Merge Main into Cut by Destination

Erabior committed:
Fixed a bug where cut all and by destination could not enable/disable handbrakes or bleed air

PrinceOfPluto commented on WaypointQueue/ManagedWaypoint.cs:
```diff
@@ -45,6 +45,19 @@ public enum TodayOrTomorrow
+        public enum UncoupleMode
+        {
+            All = 0,
+            ByCount = 1,
```
`ByCount` will need to be enum value 0 so that this is properly backwards compatible with waypoints players already have saved

PrinceOfPluto commented on WaypointQueue/ManagedWaypoint.cs:
```diff
@@ -45,6 +45,19 @@ public enum TodayOrTomorrow
+        public enum UncoupleAllDirection
+        {
+            Aft = 0,
+            Fore = 1
```
We should stick to using front and rear which better corresponds to how the base game's code refers to car direction

PrinceOfPluto replied:
I'd prefer to use Front and Rear here so it is more easily understood as reference to the car ends

PrinceOfPluto commented on WaypointQueue/ManagedWaypoint.cs:
```diff
@@ -164,7 +202,7 @@ public void Load()
 
         public bool TryResolveLocomotive(out Car loco)
         {
-            // loco is null if false
```
Could you revert this change to keep the existing comment?

PrinceOfPluto reviewed (changes requested):
Submitting a partial review with some requested changes. The big one is that a lot of the changes added into WaypointResolver are very difficult to follow and understand, particularly in ResolveUncouplingOrders

Erabior committed:
Refactored ResolveUncouplingOrders
Added 'Wait Before Cut' functionality to Uncouple Modes 'All' and 'By Destination'
Added 'Do Not Split Loco/Tender Pairs' to Uncouple By Count
Fixed mistakes made during merge

Erabior committed:
Refactored ResolveUncouplingOrders
Added 'Wait Before Cut' functionaloty to Uncouple Modes 'All' and 'By Destination'
Added 'Do Not Split Loco/Tender Pairs' to Uncouple By Count
Fixed mistakes made during merge

Erabior committed:
Merge branch 'Cut-by-destination-2' of https://github.com/PrinceOfPluto/WaypointQueue into Cut-by-destination-2

Erabior committed:
Added cut from end feature to cut by destination
Re-worded controls for cut by destination to better fit their function
Added tooltips to cut by destination fields

Erabior committed:
Added the ability to uncouple by destination and uncouple all as a post couple cut
